src/v/cloud_topics/reconciler/reconciler.cc

```diff
@@ -177,94 +169,178 @@ ss::future<> reconciler::reconciliation_loop() {
+    auto object_id = l1::create_object_id();
+    auto upload_fut = co_await ss::coroutine::as_future(
+      _l1_io->put_object(object_id, staging_file.get(), &_as));
+    co_await staging_file->remove(); // Always.
+    if (upload_fut.failed()) {
+        vlog(
+          lg.error,
+          "Exception uploading L1 object {}: {}",
+          object_id,
+          upload_fut.get_exception());
         co_return;
     }
+    auto upload_result = upload_fut.get();
+    if (!upload_result.has_value()) {
+        vlog(
+          lg.warn,
+          "Failed to upload L1 object: {}",
+          static_cast<int>(upload_result.error()));
+        co_return;
+    }
+    vlog(lg.debug, "Successfully uploaded L1 object {}", object_id);
 
-    // commit for each partition represented in the uploaded object
-    for (const auto& range : object->ranges) {
-        co_await commit_object(path, range);
+    // Commit object.
+    // TODO: This looks very different with the metastore.
+    for (const auto& partition_info : object->partitions) {
+        co_await commit_object(object_id, partition_info);
     }
 }
 
-ss::future<std::optional<reconciler::object>> reconciler::build_object() {
-    // light-weight copy for stable iteration
+ss::future<std::optional<reconciler::built_object>> reconciler::build_object(
+  l1::object_builder* builder, l1::staging_file* staging_file) {
+    // Copy the leader partition information in case of
+    // mid-reconciliation unregistration.
     std::vector<attached_partition> partitions;
     for (const auto& p : _partitions) {
         if (p.second->partition->is_leader()) {
             partitions.push_back(p.second);
         }
     }
 
-    // avoid starving partitions
+    if (partitions.empty()) {
+        vlog(lg.debug, "No leader partitions to reconcile");
+        co_return std::nullopt;
+    }
+
+    // Shuffle to avoid starving partitions.
+    // TODO: Investigate how to divide work between partitions with
+    // different throughput.
     std::shuffle(
       partitions.begin(), partitions.end(), random_generators::internal::gen);
 
-    object object;
+    built_object result;
     auto size_budget = max_object_size;
     for (const auto& partition : partitions) {
+        vlog(
+          lg.debug,
+          "Processing partition {} with LRO {}",
+          partition->partition->ntp(),
+          partition->lro);
+
         frontend fe(partition->partition, _data_plane);
         auto reader = co_await make_reader(&fe, partition->lro, size_budget);
-        auto range = co_await std::move(reader).consume(
-          range_batch_consumer{}, model::no_timeout);
-        if (range.has_value()) {
-            object.add(std::move(*range), partition);
-            size_budget -= std::min(object.data.size_bytes(), size_budget);
+        auto tidp = ntp_to_topic_id_partition(partition->partition->ntp());
+        if (!tidp.has_value()) {
+            vlog(
+              lg.warn,
```

Comment:
Let's mark this as an error, there should always be this mapping for cloud topics.